src/libraries/System.Private.CoreLib/src/System/Buffers/SharedArrayPool.cs

```diff
@@ -513,4 +522,64 @@ public ThreadLocalArray(T[] array)
             }
         }
     }
+
+    internal static class SharedArrayPoolStatics
+    {
+        /// <summary>Gets the maximum number of partitions to shard arrays into.</summary>
+        /// <remarks>Defaults to int.MaxValue.  Whatever value is returned will end up being clamped to <see cref="Environment.ProcessorCount"/>.</remarks>
+        internal static int GetPartitionCount()
+        {
+            int partitionCount = TryGetInt32EnvironmentVariable("DOTNET_SYSTEM_BUFFERS_SHAREDARRAYPOOL_MAXPARTITIONCOUNT", out int result) && result > 0 ?
+                result :
+                int.MaxValue; // no limit other than processor count
+            return Math.Min(partitionCount, Environment.ProcessorCount);
+        }
+
+        /// <summary>Gets the maximum number of arrays of a given size allowed to be cached per partition.</summary>
+        /// <returns>Defaults to 8. This does not factor in or impact the number of arrays cached per thread in TLS (currently only 1).</returns>
+        internal static int GetMaxArraysPerPartition()
+        {
+            return TryGetInt32EnvironmentVariable("DOTNET_SYSTEM_BUFFERS_SHAREDARRAYPOOL_MAXARRAYSPERPARTITION", out int result) && result > 0 ?
+                result :
+                8; // arbitrary limit
+        }
+
+        /// <summary>Look up an environment variable and try to parse it as an Int32.</summary>
+        /// <remarks>This avoids using anything that might in turn recursively use the ArrayPool.</remarks>
+        internal static bool TryGetInt32EnvironmentVariable(string variable, out int result)
+        {
+            // Avoid globalization stack, as it might in turn be using ArrayPool.
+
+            Span<char> value = stackalloc char[32]; // max integer length is 10; allow some extra spaces
+            uint length = Interop.Kernel32.GetEnvironmentVariable(variable, ref MemoryMarshal.GetReference(value), (uint)value.Length);
```

Comment:
Hmm, I could be missing something but how will this work on NativeAOT Linux/macOS? (ie. anything non-Windows without miniPAL)

Reply:
Grr, it won't.  I forget the PAL doesn't apply there.  I'll try again.